File: src/main/java/pl/edu/pw/ee/catering_backend/orders/application/OrdersController.java
package pl.edu.pw.ee.catering_backend.orders.application;

import com.fasterxml.jackson.databind.json.JsonMapper;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import pl.edu.pw.ee.catering_backend.authentication.application.services.JwtService;
import pl.edu.pw.ee.catering_backend.orders.comms.IManagerOrderingData;
import pl.edu.pw.ee.catering_backend.orders.comms.dtos.AddOrderDTO;
import pl.edu.pw.ee.catering_backend.orders.comms.dtos.OrderDto;
import pl.edu.pw.ee.catering_backend.orders.domain.OrdersService;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/orders")
@RequiredArgsConstructor
public class OrdersController implements IManagerOrderingData {

    private final JwtService jwtService;
    private final OrdersService ordersService;

    @Override
    @PostMapping("/create")
    @ApiResponses(
            value = {
                    @ApiResponse(responseCode = "201", description = "Order created"),
                    @ApiResponse(responseCode = "400", description = "Invalid input"),
                    @ApiResponse(responseCode = "401", description = "Unauthorized"),
                    @ApiResponse(responseCode = "409", description = "Order already exists")
            }
    )
    public ResponseEntity<Map<String, String>> addOrder(
            @RequestHeader("Authorization") String token,
            @RequestBody AddOrderDTO dto
    ) {
        var login = jwtService.extractLogin(token);
        var order = ordersService.addOrder(
                login,
                dto
        );
        JsonMapper jsonMapper = new JsonMapper();
        Map<String, String> orderMap = Map.of(
                "orderId", order.getOrderId().toString()
        );
        jsonMapper.valueToTree(orderMap);
        return ResponseEntity.ok().body(orderMap);
    }

    @GetMapping
    @ApiResponses(
            value = {
                    @ApiResponse(responseCode = "200", description = "Orders found"),
                    @ApiResponse(responseCode = "400", description = "Invalid input"),
                    @ApiResponse(responseCode = "404", description = "Orders not found")
            }
    )
    @Override
    public ResponseEntity<List<OrderDto>> getOrders(
            @RequestHeader("Authorization") String token
    ) {
        var login = jwtService.extractLogin(token);
        var orders = ordersService.getOrders(login);
        if (orders.isEmpty()) {
            return ResponseEntity.notFound().build();
        }
        return ResponseEntity.ok(orders);
    }

    @GetMapping("/{id}")
    @Override
    @ApiResponses(
            value = {
                    @ApiResponse(responseCode = "200", description = "Order found"),
                    @ApiResponse(responseCode = "400", description = "Invalid input"),
                    @ApiResponse(responseCode = "404", description = "Order not found")
            }
    )
    public ResponseEntity<OrderDto> getOrderById(@PathVariable String id) {
        var uuid = java.util.UUID.fromString(id);
        var order = ordersService.getOrderById(uuid);
        if (order == null) {
            return ResponseEntity.notFound().build();
        }
        return ResponseEntity.ok(order);
    }
}

File: src/main/java/pl/edu/pw/ee/catering_backend/orders/application/OrdersController.kt
package pl.edu.pw.ee.catering_backend.orders.application

import io.swagger.v3.oas.annotations.responses.ApiResponse
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import pl.edu.pw.ee.catering_backend.orders.domain.CreateOrderUseCase
import java.util.logging.Logger
import kotlin.math.log

@RestController
@RequestMapping("/orders")
class OrdersController {
    @Autowired
    private lateinit var createOrderUseCase: CreateOrderUseCase

    private val logger = Logger.getLogger(OrdersController::class.java.name)

    @PostMapping("/create")
    @ApiResponse(description = "Creates order")
    @ApiResponse(responseCode = "200", description = "Order created")
    @ApiResponse(responseCode = "400", description = "Bad request")
    @ApiResponse(responseCode = "500", description = "Internal server error")
    @ApiResponse(responseCode = "401", description = "Unauthorized")
    fun createOrder(@RequestBody request: OrderCreationRequest): ResponseEntity<OrderCreationResponse> {
        if (!areAllArgumentsValid(request)) {
            logger.warning("Invalid arguments")
            return ResponseEntity.badRequest().body(OrderCreationResponse(error = "Invalid arguments"))
        }
        logger.info("Creating order")

        val result = createOrderUseCase(
            request.clientLogin,
            request.mealsIds,
            request.deliveryAddress,
            request.deliveryMethod,
            request.status
        )

        return result.fold(
            { ResponseEntity.ok(OrderCreationResponse(orderId = it)) },
            {
                ResponseEntity.badRequest().body(OrderCreationResponse(error = it.message ?: "Unknown error"))
            }
        )
    }

    private fun areAllArgumentsValid(request: OrderCreationRequest): Boolean {
        return request.clientLogin.isNotEmpty() && request.mealsIds.isNotEmpty() && request.deliveryAddress.isNotEmpty() && request.deliveryMethod.isNotEmpty() && request.status.isNotEmpty()
    }
}

data class OrderCreationResponse(
    val orderId: String? = null,
    val error: String? = null,
)
